<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="Sebastian Raschka">
        <link rel="canonical" href="http://rasbt.github.io/biopandas/tutorials/test/">
        <link rel="shortcut icon" href="../../favicon.ico">
        
        <title>Working with PDB Structures in DataFrames - BioPandas</title>
    <link href="../../css/bootstrap-3.3.7.min.css" rel="stylesheet">
    <link href="../../css/font-awesome.min.css" rel="stylesheet">
    <link href="../../css/base.css" rel="stylesheet">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
    <link href="../../extra.css" rel="stylesheet">
    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
        <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
        <script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
    <![endif]-->

    <script src="../../js/jquery-1.10.2.min.js" defer></script>
    <script src="../../js/bootstrap-3.3.7.min.js" defer></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-38457794-3', 'rasbt.github.io/biopandas');
            ga('send', 'pageview');
        </script> 
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
            <div class="container">

                <!-- Collapsed navigation -->
                <div class="navbar-header">
                    <!-- Expander button -->
                    <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                        <span class="sr-only">Toggle navigation</span>
                        <span class="icon-bar"></span>
                        <span class="icon-bar"></span>
                        <span class="icon-bar"></span>
                    </button>
                    <a class="navbar-brand" href="../..">BioPandas</a>
                </div>

                <!-- Expanded navigation -->
                <div class="navbar-collapse collapse">
                        <!-- Main navigation -->
                        <ul class="nav navbar-nav">
                            <li >
                                <a href="../..">Home</a>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="dropdown-toggle" data-toggle="dropdown">Tutorials <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li >
    <a href="../Working_with_PDB_Structures_in_DataFrames/">Working with PDB Structures in DataFrames</a>
</li>
                                    
<li >
    <a href="../Working_with_MOL2_Structures_in_DataFrames/">Working with MOL2 Structures in DataFrames</a>
</li>
                                </ul>
                            </li>
                            <li class="dropdown">
                                <a href="#" class="dropdown-toggle" data-toggle="dropdown">API <b class="caret"></b></a>
                                <ul class="dropdown-menu">
                                    
<li >
    <a href="../../api_subpackages/biopandas.pdb/">biopandas.pdb</a>
</li>
                                    
<li >
    <a href="../../api_subpackages/biopandas.mol2/">biopandas.mol2</a>
</li>
                                </ul>
                            </li>
                            <li >
                                <a href="../../CHANGELOG/">Changelog</a>
                            </li>
                            <li >
                                <a href="../../installation/">Installation</a>
                            </li>
                            <li >
                                <a href="../../CONTRIBUTING/">Contributing</a>
                            </li>
                            <li >
                                <a href="../../citing/">Citing</a>
                            </li>
                        </ul>

                    <ul class="nav navbar-nav navbar-right">
                        <li>
                            <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                                <i class="fa fa-search"></i> Search
                            </a>
                        </li>
                            <li>
                                <a href="https://github.com/rasbt/biopandas/edit/master/docs/tutorials/test.md"><i class="fa fa-github"></i> Edit on GitHub</a>
                            </li>
                    </ul>
                </div>
            </div>
        </div>

        <div class="container">
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="main active"><a href="#working-with-pdb-structures-in-dataframes">Working with PDB Structures in DataFrames</a></li>
            <li><a href="#loading-pdb-files">Loading PDB Files</a></li>
            <li><a href="#looking-at-pdbs-in-dataframes">Looking at PDBs in DataFrames</a></li>
            <li><a href="#working-with-pdb-dataframes">Working with PDB DataFrames</a></li>
            <li><a href="#plotting">Plotting</a></li>
            <li><a href="#computing-the-root-mean-square-deviation">Computing the Root Mean Square Deviation</a></li>
            <li><a href="#filtering-pdbs-by-distance">Filtering PDBs by Distance</a></li>
            <li><a href="#converting-amino-acid-codes-from-3-to-1-letter-codes">Converting Amino Acid codes from 3- to 1-letter codes</a></li>
            <li><a href="#wrapping-it-up-saving-pdb-structures">Wrapping it up - Saving PDB structures</a></li>
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="working-with-pdb-structures-in-dataframes">Working with PDB Structures in DataFrames</h1>
<h2 id="loading-pdb-files">Loading PDB Files</h2>
<p>There are 2 1/2 ways to load a PDB structure into a <code>PandasPdb</code> object.</p>
<h4 id="1">1</h4>
<p>PDB files can be directly fetched from The Protein Data Bank at <a href="http://www.rcsb.org">http://www.rcsb.org</a> via its unique 4-letter after initializing a new <a href="../api_subpackages/biopandas.pdb#pandaspdb"><code>PandasPdb</code></a> object and calling the <a href="../api_subpackages/biopandas.pdb#pandaspdbfetch_pdb"><code>fetch_pdb</code></a> method:</p>
<pre><code class="python">from biopandas.pdb import PandasPdb

# Initialize a new PandasPdb object
# and fetch the PDB file from rcsb.org
ppdb = PandasPdb().fetch_pdb('3eiy')
</code></pre>

<h4 id="2-a">2 a)</h4>
<p>Alternatively, we can load PDB files from local directories as regular PDB files using <a href="../api_subpackages/biopandas.pdb#pandaspdbread_pdb"><code>read_pdb</code></a>:</p>
<pre><code class="python">ppdb.read_pdb('./data/3eiy.pdb')
</code></pre>

<pre><code>&lt;biopandas.pdb.pandas_pdb.PandasPdb at 0x11912b710&gt;
</code></pre>
<p>[File link: <a href="https://raw.githubusercontent.com/rasbt/biopandas/master/docs/sources/tutorials/data/3eiy.pdb">3eiy.pdb</a>]</p>
<h4 id="2-b">2 b)</h4>
<p>Or, we can load them from gzip archives like so (note that the file must end with a '.gz' suffix in order to be recognized as a gzip file):</p>
<pre><code class="python">ppdb.read_pdb('./data/3eiy.pdb.gz')
</code></pre>

<pre><code>&lt;biopandas.pdb.pandas_pdb.PandasPdb at 0x11912b710&gt;
</code></pre>
<p>[File link: <a href="https://github.com/rasbt/biopandas/blob/master/docs/sources/tutorials/data/3eiy.pdb.gz?raw=true">3eiy.pdb.gz</a>]</p>
<p>After the file was succesfully loaded, we have access to the following attributes:</p>
<pre><code class="python">print('PDB Code: %s' % ppdb.code)
print('PDB Header Line: %s' % ppdb.header)
print('\nRaw PDB file contents:\n\n%s\n...' % ppdb.pdb_text[:1000])
</code></pre>

<pre><code>PDB Code: 3eiy
PDB Header Line:     HYDROLASE                               17-SEP-08   3EIY

Raw PDB file contents:

HEADER    HYDROLASE                               17-SEP-08   3EIY              
TITLE     CRYSTAL STRUCTURE OF INORGANIC PYROPHOSPHATASE FROM BURKHOLDERIA      
TITLE    2 PSEUDOMALLEI WITH BOUND PYROPHOSPHATE                                
COMPND    MOL_ID: 1;                                                            
COMPND   2 MOLECULE: INORGANIC PYROPHOSPHATASE;                                 
COMPND   3 CHAIN: A;                                                            
COMPND   4 EC: 3.6.1.1;                                                         
COMPND   5 ENGINEERED: YES                                                      
SOURCE    MOL_ID: 1;                                                            
SOURCE   2 ORGANISM_SCIENTIFIC: BURKHOLDERIA PSEUDOMALLEI 1710B;                
SOURCE   3 ORGANISM_TAXID: 320372;                                              
SOURCE   4 GENE: PPA, BURPS1710B_1237;                                          
SOURCE   5 EXPRESSION_SYSTEM
...
</code></pre>
<p>The most interesting / useful attribute is the <a href="../api_subpackages/biopandas.pdb#pandaspdbdf"><code>PandasPdb.df</code></a> DataFrame dictionary though, which gives us access to the PDB files as pandas DataFrames. Let's print the first 3 lines from the <code>ATOM</code> coordinate section to see how it looks like:</p>
<pre><code class="python">ppdb.df['ATOM'].head(3)
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>atom_number</th>
      <th>blank_1</th>
      <th>atom_name</th>
      <th>...</th>
      <th>segment_id</th>
      <th>element_symbol</th>
      <th>charge</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ATOM</td>
      <td>1</td>
      <td></td>
      <td>N</td>
      <td>...</td>
      <td></td>
      <td>N</td>
      <td>NaN</td>
      <td>609</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ATOM</td>
      <td>2</td>
      <td></td>
      <td>CA</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>610</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ATOM</td>
      <td>3</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>611</td>
    </tr>
  </tbody>
</table>
<p>3 rows × 21 columns</p>
</div>

<p>But more on that in the next section.</p>
<h2 id="looking-at-pdbs-in-dataframes">Looking at PDBs in DataFrames</h2>
<p>PDB files are parsed according to the <a href="http://www.rcsb.org/pdb/static.do?p=file_formats/pdb/index.html">PDB file format description</a>. More specifically, BioPandas reads the columns of the ATOM and HETATM sections as shown in the following excerpt from <a href="http://deposit.rcsb.org/adit/docs/pdb_atom_format.html#ATOM">http://deposit.rcsb.org/adit/docs/pdb_atom_format.html#ATOM</a>.</p>
<table>
<thead>
<tr>
<th>COLUMNS</th>
<th>DATA TYPE</th>
<th>CONTENTS</th>
<th>biopandas column name</th>
</tr>
</thead>
<tbody>
<tr>
<td>1 - 6</td>
<td>Record name</td>
<td>"ATOM"</td>
<td>record_name</td>
</tr>
<tr>
<td>7 - 11</td>
<td>Integer</td>
<td>Atom serial number.</td>
<td>atom_number</td>
</tr>
<tr>
<td>12</td>
<td></td>
<td></td>
<td>blank_1</td>
</tr>
<tr>
<td>13 - 16</td>
<td>Atom</td>
<td>Atom name.</td>
<td>atom_name</td>
</tr>
<tr>
<td>17</td>
<td>Character</td>
<td>Alternate location indicator.</td>
<td>alt_loc</td>
</tr>
<tr>
<td>18 - 20</td>
<td>Residue name</td>
<td>Residue name.</td>
<td>residue_name</td>
</tr>
<tr>
<td>21</td>
<td></td>
<td></td>
<td>blank_2</td>
</tr>
<tr>
<td>22</td>
<td>Character</td>
<td>Chain identifier.</td>
<td>chain_id</td>
</tr>
<tr>
<td>23 - 26</td>
<td>Integer</td>
<td>Residue sequence number.</td>
<td>residue_number</td>
</tr>
<tr>
<td>27</td>
<td>AChar</td>
<td>Code for insertion of residues.</td>
<td>insertion</td>
</tr>
<tr>
<td>28 - 30</td>
<td></td>
<td></td>
<td>blank_3</td>
</tr>
<tr>
<td>31 - 38</td>
<td>Real(8.3)</td>
<td>Orthogonal coordinates for X in Angstroms.</td>
<td>x_coord</td>
</tr>
<tr>
<td>39 - 46</td>
<td>Real(8.3)</td>
<td>Orthogonal coordinates for Y in Angstroms.</td>
<td>y_coord</td>
</tr>
<tr>
<td>47 - 54</td>
<td>Real(8.3)</td>
<td>Orthogonal coordinates for Z in Angstroms.</td>
<td>z_coord</td>
</tr>
<tr>
<td>55 - 60</td>
<td>Real(6.2)</td>
<td>Occupancy.</td>
<td>occupancy</td>
</tr>
<tr>
<td>61 - 66</td>
<td>Real(6.2)</td>
<td>Temperature factor (Default = 0.0).</td>
<td>bfactor</td>
</tr>
<tr>
<td>67-72</td>
<td></td>
<td></td>
<td>blank_4</td>
</tr>
<tr>
<td>73 - 76</td>
<td>LString(4)</td>
<td>Segment identifier, left-justified.</td>
<td>segment_id</td>
</tr>
<tr>
<td>77 - 78</td>
<td>LString(2)</td>
<td>Element symbol, right-justified.</td>
<td>element_symbol</td>
</tr>
<tr>
<td>79 - 80</td>
<td>LString(2)</td>
<td>Charge on the atom.</td>
<td>charge</td>
</tr>
</tbody>
</table>
<p>Below is an example of how this would look like in an actual PDB file:</p>
<pre><code>Example: 
         1         2         3         4         5         6         7         8
12345678901234567890123456789012345678901234567890123456789012345678901234567890
ATOM    145  N   VAL A  25      32.433  16.336  57.540  1.00 11.92      A1   N
ATOM    146  CA  VAL A  25      31.132  16.439  58.160  1.00 11.85      A1   C
ATOM    147  C   VAL A  25      30.447  15.105  58.363  1.00 12.34      A1   C
ATOM    148  O   VAL A  25      29.520  15.059  59.174  1.00 15.65      A1   O
ATOM    149  CB AVAL A  25      30.385  17.437  57.230  0.28 13.88      A1   C
ATOM    150  CB BVAL A  25      30.166  17.399  57.373  0.72 15.41      A1   C
ATOM    151  CG1AVAL A  25      28.870  17.401  57.336  0.28 12.64      A1   C
ATOM    152  CG1BVAL A  25      30.805  18.788  57.449  0.72 15.11      A1   C
ATOM    153  CG2AVAL A  25      30.835  18.826  57.661  0.28 13.58      A1   C
ATOM    154  CG2BVAL A  25      29.909  16.996  55.922  0.72 13.25      A1   C
</code></pre>
<p>After loading a PDB file from rcsb.org or our local drive, the <a href="../api_subpackages/biopandas.pdb/#pandaspdbdf"><code>PandasPdb.df</code></a> attribute should contain the following 4 DataFrame objects:</p>
<pre><code class="python">from biopandas.pdb import PandasPdb
ppdb = PandasPdb()
ppdb.read_pdb('./data/3eiy.pdb')
ppdb.df.keys()
</code></pre>

<pre><code>dict_keys(['ATOM', 'HETATM', 'ANISOU', 'OTHERS'])
</code></pre>
<p>[File link: <a href="https://raw.githubusercontent.com/rasbt/biopandas/master/docs/sources/tutorials/data/3eiy.pdb">3eiy.pdb</a>]</p>
<ul>
<li>'ATOM': contains the entries from the ATOM coordinate section</li>
<li>'ATOM':  ... entries from the "HETATM" coordinate section    </li>
<li>'ANISOU': ... entries from the "ANISOU" coordinate section </li>
<li>'OTHERS': Everything else that is <em>not</em> a 'ATOM', 'HETATM', or 'ANISOU' entry</li>
</ul>
<p><img alt="" src="../img/df_dict.jpg" /></p>
<p>The columns of the 'HETATM' DataFrame are indentical to the 'ATOM' DataFrame that we've seen earlier:</p>
<pre><code class="python">ppdb.df['HETATM'].head(2)
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>atom_number</th>
      <th>blank_1</th>
      <th>atom_name</th>
      <th>...</th>
      <th>segment_id</th>
      <th>element_symbol</th>
      <th>charge</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>HETATM</td>
      <td>1332</td>
      <td></td>
      <td>K</td>
      <td>...</td>
      <td></td>
      <td>K</td>
      <td>NaN</td>
      <td>1940</td>
    </tr>
    <tr>
      <th>1</th>
      <td>HETATM</td>
      <td>1333</td>
      <td></td>
      <td>NA</td>
      <td>...</td>
      <td></td>
      <td>NA</td>
      <td>NaN</td>
      <td>1941</td>
    </tr>
  </tbody>
</table>
<p>2 rows × 21 columns</p>
</div>

<p><br></p>
<p>Note that "ANISOU" entries are handled a bit differently as specified at <a href="http://deposit.rcsb.org/adit/docs/pdb_atom_format.html#ATOM">http://deposit.rcsb.org/adit/docs/pdb_atom_format.html#ATOM</a>.</p>
<pre><code class="python">ppdb.df['ANISOU'].head(2)
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>atom_number</th>
      <th>blank_1</th>
      <th>atom_name</th>
      <th>...</th>
      <th>blank_4</th>
      <th>element_symbol</th>
      <th>charge</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
  </tbody>
</table>
<p>0 rows × 21 columns</p>
</div>

<p>Not every PDB file contains ANISOU entries (similarly, some PDB files may only contain HETATM or ATOM entries). If records are basent, the DataFrame will be empty as show above.</p>
<pre><code class="python">ppdb.df['ANISOU'].empty
</code></pre>

<pre><code>True
</code></pre>
<p>Since the DataFrames are fairly wide, let's us take a look at the columns by accessing the DataFrame's <code>column</code> attribute:</p>
<pre><code class="python">ppdb.df['ANISOU'].columns
</code></pre>

<pre><code>Index(['record_name', 'atom_number', 'blank_1', 'atom_name', 'alt_loc', 'residue_name', 'blank_2', 'chain_id', 'residue_number', 'insertion', 'blank_3', 'U(1,1)', 'U(2,2)', 'U(3,3)', 'U(1,2)', 'U(1,3)', 'U(2,3)', 'blank_4', 'element_symbol', 'charge', 'line_idx'], dtype='object')
</code></pre>
<p>ANISOU records are very similar to ATOM/HETATM records. In fact, the columns 7 - 27 and 73 - 80 are identical to their corresponding ATOM/HETATM records, which means that the 'ANISOU' DataFrame doesn't have the following entries:</p>
<pre><code class="python">set(ppdb.df['ATOM'].columns).difference(set(ppdb.df['ANISOU'].columns))
</code></pre>

<pre><code>{'b_factor', 'occupancy', 'segment_id', 'x_coord', 'y_coord', 'z_coord'}
</code></pre>
<p>Instead, the "ANISOU" DataFrame contains the anisotropic temperature factors "U(-,-)" -- note that these are scaled by a factor of <script type="math/tex">10^4</script> (<script type="math/tex">\text{Angstroms}^2</script>) by convention.</p>
<pre><code class="python">set(ppdb.df['ANISOU'].columns).difference(set(ppdb.df['ATOM'].columns))
</code></pre>

<pre><code>{'U(1,1)', 'U(1,2)', 'U(1,3)', 'U(2,2)', 'U(2,3)', 'U(3,3)'}
</code></pre>
<p><br>
<br></p>
<p>Ah, another interesting thing to mention is that the columns already come with the types you'd expect (where <code>object</code> essentially "means" <code>str</code> here):</p>
<pre><code class="python">ppdb.df['ATOM'].dtypes
</code></pre>

<pre><code>record_name        object
atom_number         int64
blank_1            object
atom_name          object
alt_loc            object
residue_name       object
blank_2            object
chain_id           object
residue_number      int64
insertion          object
blank_3            object
x_coord           float64
y_coord           float64
z_coord           float64
occupancy         float64
b_factor          float64
blank_4            object
segment_id         object
element_symbol     object
charge            float64
line_idx            int64
dtype: object
</code></pre>
<p><br></p>
<p>Typically, all good things come in threes, however, there is a 4th DataFrame, an'OTHER' DataFrame, which contains everything that wasn't parsed as 'ATOM', 'HETATM', or 'ANISOU' coordinate section:</p>
<pre><code class="python">ppdb.df['OTHERS'].head(5)
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>entry</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>HEADER</td>
      <td>HYDROLASE                               17...</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TITLE</td>
      <td>CRYSTAL STRUCTURE OF INORGANIC PYROPHOSPHA...</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>TITLE</td>
      <td>2 PSEUDOMALLEI WITH BOUND PYROPHOSPHATE</td>
      <td>2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>COMPND</td>
      <td>MOL_ID: 1;</td>
      <td>3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>COMPND</td>
      <td>2 MOLECULE: INORGANIC PYROPHOSPHATASE;</td>
      <td>4</td>
    </tr>
  </tbody>
</table>
</div>

<p>Although these 'OTHER' entries are typically less useful for structure-related computations, you may still want to take a look at them to get a short summary of the PDB structure and learn about it's potential quirks and gotchas (typically listed in the REMARKs section). Lastly, the "OTHERS" DataFrame comes in handy if we want to reconstruct the structure as PDB file as we will see later (note the <code>line_idx</code> columns in all of the DataFrames).</p>
<h2 id="working-with-pdb-dataframes">Working with PDB DataFrames</h2>
<p>In the previous sections, we've seen how to load PDB structures into DataFrames, and how to access them. Now, let's talk about manipulating PDB files in DataFrames.</p>
<pre><code class="python">from biopandas.pdb import PandasPdb
ppdb = PandasPdb()
ppdb.read_pdb('./data/3eiy.pdb.gz')
ppdb.df['ATOM'].head()
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>atom_number</th>
      <th>blank_1</th>
      <th>atom_name</th>
      <th>...</th>
      <th>segment_id</th>
      <th>element_symbol</th>
      <th>charge</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ATOM</td>
      <td>1</td>
      <td></td>
      <td>N</td>
      <td>...</td>
      <td></td>
      <td>N</td>
      <td>NaN</td>
      <td>609</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ATOM</td>
      <td>2</td>
      <td></td>
      <td>CA</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>610</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ATOM</td>
      <td>3</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>611</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ATOM</td>
      <td>4</td>
      <td></td>
      <td>O</td>
      <td>...</td>
      <td></td>
      <td>O</td>
      <td>NaN</td>
      <td>612</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ATOM</td>
      <td>5</td>
      <td></td>
      <td>CB</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>613</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21 columns</p>
</div>

<p>[File link: <a href="https://github.com/rasbt/biopandas/blob/master/docs/sources/tutorials/data/3eiy.pdb.gz?raw=true">3eiy.pdb.gz</a>]</p>
<p>Okay, there's actually not <em>that</em> much to say ... <br />
Once we have our PDB file in the DataFrame format, we have the whole convenience of <a href="http://pandas.pydata.org">pandas</a> right there at our fingertips.</p>
<p>For example, let's get all Proline residues:</p>
<pre><code class="python">ppdb.df['ATOM'][ppdb.df['ATOM']['residue_name'] == 'PRO'].head()
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>atom_number</th>
      <th>blank_1</th>
      <th>atom_name</th>
      <th>...</th>
      <th>segment_id</th>
      <th>element_symbol</th>
      <th>charge</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>38</th>
      <td>ATOM</td>
      <td>39</td>
      <td></td>
      <td>N</td>
      <td>...</td>
      <td></td>
      <td>N</td>
      <td>NaN</td>
      <td>647</td>
    </tr>
    <tr>
      <th>39</th>
      <td>ATOM</td>
      <td>40</td>
      <td></td>
      <td>CA</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>648</td>
    </tr>
    <tr>
      <th>40</th>
      <td>ATOM</td>
      <td>41</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>649</td>
    </tr>
    <tr>
      <th>41</th>
      <td>ATOM</td>
      <td>42</td>
      <td></td>
      <td>O</td>
      <td>...</td>
      <td></td>
      <td>O</td>
      <td>NaN</td>
      <td>650</td>
    </tr>
    <tr>
      <th>42</th>
      <td>ATOM</td>
      <td>43</td>
      <td></td>
      <td>CB</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>651</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21 columns</p>
</div>

<p>Or main chain atoms:</p>
<pre><code class="python">ppdb.df['ATOM'][ppdb.df['ATOM']['atom_name'] == 'C'].head()
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>atom_number</th>
      <th>blank_1</th>
      <th>atom_name</th>
      <th>...</th>
      <th>segment_id</th>
      <th>element_symbol</th>
      <th>charge</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>ATOM</td>
      <td>3</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>611</td>
    </tr>
    <tr>
      <th>8</th>
      <td>ATOM</td>
      <td>9</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>617</td>
    </tr>
    <tr>
      <th>19</th>
      <td>ATOM</td>
      <td>20</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>628</td>
    </tr>
    <tr>
      <th>25</th>
      <td>ATOM</td>
      <td>26</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>634</td>
    </tr>
    <tr>
      <th>33</th>
      <td>ATOM</td>
      <td>34</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>642</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21 columns</p>
</div>

<p>It's also easy to strip our coordinate section from hydrogen atoms if there are any ...</p>
<pre><code class="python">ppdb.df['ATOM'][ppdb.df['ATOM']['element_symbol'] != 'H'].head()
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>atom_number</th>
      <th>blank_1</th>
      <th>atom_name</th>
      <th>...</th>
      <th>segment_id</th>
      <th>element_symbol</th>
      <th>charge</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ATOM</td>
      <td>1</td>
      <td></td>
      <td>N</td>
      <td>...</td>
      <td></td>
      <td>N</td>
      <td>NaN</td>
      <td>609</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ATOM</td>
      <td>2</td>
      <td></td>
      <td>CA</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>610</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ATOM</td>
      <td>3</td>
      <td></td>
      <td>C</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>611</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ATOM</td>
      <td>4</td>
      <td></td>
      <td>O</td>
      <td>...</td>
      <td></td>
      <td>O</td>
      <td>NaN</td>
      <td>612</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ATOM</td>
      <td>5</td>
      <td></td>
      <td>CB</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>613</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21 columns</p>
</div>

<p>Or, let's compute the average temperature factor of our protein main chain:</p>
<pre><code class="python">mainchain = ppdb.df['ATOM'][(ppdb.df['ATOM']['atom_name'] == 'C') | 
                            (ppdb.df['ATOM']['atom_name'] == 'O') | 
                            (ppdb.df['ATOM']['atom_name'] == 'N') | 
                            (ppdb.df['ATOM']['atom_name'] == 'CA')]

bfact_mc_avg = mainchain['b_factor'].mean()
print('Average B-Factor [Main Chain]: %.2f' % bfact_mc_avg)
</code></pre>

<pre><code>Average B-Factor [Main Chain]: 28.83
</code></pre>
<h2 id="plotting">Plotting</h2>
<p>Since we are using pandas under the hood, which in turns uses matplotlib under the hood, we can produce quick summary plots of our PDB structures relatively conveniently:</p>
<pre><code class="python">from biopandas.pdb import PandasPdb
ppdb = PandasPdb().read_pdb('./data/3eiy.pdb.gz')
</code></pre>

<p>[File link: <a href="https://github.com/rasbt/biopandas/blob/master/docs/sources/tutorials/data/3eiy.pdb.gz?raw=true">3eiy.pdb.gz</a>]</p>
<pre><code class="python">%matplotlib inline
import matplotlib.pyplot as plt
from matplotlib import style
style.use('ggplot')
</code></pre>

<pre><code class="python">ppdb.df['ATOM']['b_factor'].plot(kind='hist')
plt.title('Distribution of B-Factors')
plt.xlabel('B-factor')
plt.ylabel('count')
plt.show()
</code></pre>

<p><img alt="png" src="../Working_with_PDB_Structures_in_DataFrames_files/Working_with_PDB_Structures_in_DataFrames_67_0.png" /></p>
<pre><code class="python">ppdb.df['ATOM']['b_factor'].plot(kind='line')
plt.title('B-Factors Along the Amino Acid Chain')
plt.xlabel('Residue Number')
plt.ylabel('B-factor in $A^2$')
plt.show()
</code></pre>

<p><img alt="png" src="../Working_with_PDB_Structures_in_DataFrames_files/Working_with_PDB_Structures_in_DataFrames_68_0.png" /></p>
<pre><code class="python">ppdb.df['ATOM']['element_symbol'].value_counts().plot(kind='bar')
plt.title('Distribution of Atom Types')
plt.xlabel('elements')
plt.ylabel('count')
plt.show()
</code></pre>

<p><img alt="png" src="../Working_with_PDB_Structures_in_DataFrames_files/Working_with_PDB_Structures_in_DataFrames_69_0.png" /></p>
<h2 id="computing-the-root-mean-square-deviation">Computing the Root Mean Square Deviation</h2>
<p>BioPandas also comes with certain convenience functions, for example, ...</p>
<p>The Root-mean-square deviation (RMSD) is simply a measure of the average distance between atoms of 2 protein or ligand structures. This calculation of the Cartesian error follows the equation:</p>
<p>
<script type="math/tex; mode=display">RMSD(a, b) = \sqrt{\frac{1}{n} \sum^{n}_{i=1} \big((a_{ix})^2 + (a_{iy})^2 + (a_{iz})^2 \big)} \\
= \sqrt{\frac{1}{n} \sum^{n}_{i=1} || a_i + b_i||_2^2}</script>
</p>
<p>So, assuming that the we have the following 2 conformations of a ligand molecule</p>
<p><img alt="" src="../img/ligand_rmsd.png" /></p>
<p>we can compute the RMSD as follows:</p>
<pre><code class="python">from biopandas.pdb import PandasPdb

l_1 = PandasPdb().read_pdb('./data/lig_conf_1.pdb')
l_2 = PandasPdb().read_pdb('./data/lig_conf_2.pdb')
r = PandasPdb.rmsd(l_1.df['HETATM'], l_2.df['HETATM'],
                   s=None) # all atoms, including hydrogens
print('RMSD: %.4f Angstrom' % r)
</code></pre>

<pre><code>RMSD: 2.6444 Angstrom
</code></pre>
<p>[File links: <a href="https://raw.githubusercontent.com/rasbt/biopandas/master/docs/sources/tutorials/data/lig_conf_1.pdb">lig_conf_1.pdb</a>, <a href="https://raw.githubusercontent.com/rasbt/biopandas/master/docs/sources/tutorials/data/lig_conf_2.pdb">lig_conf_2.pdb</a>]</p>
<pre><code class="python">r = PandasPdb.rmsd(l_1.df['HETATM'], l_2.df['HETATM'], 
                   s='carbon') # carbon atoms only
print('RMSD: %.4f Angstrom' % r)
</code></pre>

<pre><code>RMSD: 3.1405 Angstrom
</code></pre>
<pre><code class="python">r = PandasPdb.rmsd(l_1.df['HETATM'], l_2.df['HETATM'], 
                   s='heavy') # heavy atoms only
print('RMSD: %.4f Angstrom' % r)
</code></pre>

<pre><code>RMSD: 1.9959 Angstrom
</code></pre>
<p>Similarly, we can compute the RMSD between 2 related protein structures:</p>
<p><img alt="" src="../img/1t48_rmsd.png" /></p>
<p>The hydrogen-free RMSD:</p>
<pre><code class="python">p_1 = PandasPdb().read_pdb('./data/1t48_995.pdb')
p_2 = PandasPdb().read_pdb('./data/1t49_995.pdb')
r = PandasPdb.rmsd(p_1.df['ATOM'], p_2.df['ATOM'], s='heavy')
print('RMSD: %.4f Angstrom' % r)
</code></pre>

<pre><code>RMSD: 0.7377 Angstrom
</code></pre>
<p>Or the RMSD between the main chains only:</p>
<pre><code class="python">p_1 = PandasPdb().read_pdb('./data/1t48_995.pdb')
p_2 = PandasPdb().read_pdb('./data/1t49_995.pdb')
r = PandasPdb.rmsd(p_1.df['ATOM'], p_2.df['ATOM'], s='main chain')
print('RMSD: %.4f Angstrom' % r)
</code></pre>

<pre><code>RMSD: 0.4781 Angstrom
</code></pre>
<p><br></p>
<h2 id="filtering-pdbs-by-distance">Filtering PDBs by Distance</h2>
<p>We can use the <code>distance</code> method to compute the distance between each atom (or a subset of atoms) in our data frame and a three-dimensional reference point. For example:</p>
<pre><code class="python">p_1 = PandasPdb().read_pdb('./data/3eiy.pdb')

reference_point = (9.362, 41.410, 10.542)
distances = p_1.distance(xyz=reference_point, records=('ATOM',))
</code></pre>

<p>[File link: <a href="https://raw.githubusercontent.com/rasbt/biopandas/master/docs/sources/tutorials/data/3eiy.pdb">3eiy.pdb</a>]</p>
<p>The distance method returns a Pandas Series object:</p>
<pre><code class="python">distances.head()
</code></pre>

<pre><code>0    19.267419
1    18.306060
2    16.976934
3    16.902897
4    18.124171
dtype: float64
</code></pre>
<p>And we can use this <code>Series</code> object, for instance, to select certain atoms in our DataFrame that fall within a desired distance threshold. For example, let's select all atoms that are within 7A of our reference point: </p>
<pre><code class="python">all_within_7A = p_1.df['ATOM'][distances &lt; 7.0]
all_within_7A.tail()
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>atom_number</th>
      <th>blank_1</th>
      <th>atom_name</th>
      <th>...</th>
      <th>segment_id</th>
      <th>element_symbol</th>
      <th>charge</th>
      <th>line_idx</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>786</th>
      <td>ATOM</td>
      <td>787</td>
      <td></td>
      <td>CB</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>1395</td>
    </tr>
    <tr>
      <th>787</th>
      <td>ATOM</td>
      <td>788</td>
      <td></td>
      <td>CG</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>1396</td>
    </tr>
    <tr>
      <th>788</th>
      <td>ATOM</td>
      <td>789</td>
      <td></td>
      <td>CD1</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>1397</td>
    </tr>
    <tr>
      <th>789</th>
      <td>ATOM</td>
      <td>790</td>
      <td></td>
      <td>CD2</td>
      <td>...</td>
      <td></td>
      <td>C</td>
      <td>NaN</td>
      <td>1398</td>
    </tr>
    <tr>
      <th>790</th>
      <td>ATOM</td>
      <td>791</td>
      <td></td>
      <td>N</td>
      <td>...</td>
      <td></td>
      <td>N</td>
      <td>NaN</td>
      <td>1399</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 21 columns</p>
</div>

<p>Visualized in PyMOL, this subset (yellow surface) would look as follows:</p>
<p><img alt="" src="../img/3eiy_7a.png" /></p>
<h2 id="converting-amino-acid-codes-from-3-to-1-letter-codes">Converting Amino Acid codes from 3- to 1-letter codes</h2>
<p>Residues in the <code>residue_name</code> field can be converted into 1-letter amino acid codes, which may be useful for further sequence analysis, for example, pair-wise or multiple sequence alignments:</p>
<pre><code class="python">from biopandas.pdb import PandasPdb
ppdb = PandasPdb().fetch_pdb('5mtn')
sequence = ppdb.amino3to1()
sequence.tail()
</code></pre>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>chain_id</th>
      <th>residue_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1378</th>
      <td>B</td>
      <td>I</td>
    </tr>
    <tr>
      <th>1386</th>
      <td>B</td>
      <td>N</td>
    </tr>
    <tr>
      <th>1394</th>
      <td>B</td>
      <td>Y</td>
    </tr>
    <tr>
      <th>1406</th>
      <td>B</td>
      <td>R</td>
    </tr>
    <tr>
      <th>1417</th>
      <td>B</td>
      <td>T</td>
    </tr>
  </tbody>
</table>
</div>

<p>As shown above, the <code>amino3to1</code> method returns a <code>DataFrame</code> containing the <code>chain_id</code> and <code>residue_name</code> of the translated 1-letter amino acids. If you like to work with the sequence as a Python list of string characters, you could do the following:</p>
<pre><code class="python">sequence_list = list(sequence.loc[sequence['chain_id'] == 'A', 'residue_name'])
sequence_list[-5:] # last 5 residues of chain A
</code></pre>

<pre><code>['V', 'R', 'H', 'Y', 'T']
</code></pre>
<p>And if you prefer to work with the sequence as a string, you can use the <code>join</code> method: </p>
<pre><code class="python">''.join(sequence.loc[sequence['chain_id'] == 'A', 'residue_name'])
</code></pre>

<pre><code>'SLEPEPWFFKNLSRKDAERQLLAPGNTHGSFLIRESESTAGSFSLSVRDFDQGEVVKHYKIRNLDNGGFYISPRITFPGLHELVRHYT'
</code></pre>
<p>To iterate over the sequences of multi-chain proteins, you can use the <code>unique</code> method as shown below:</p>
<pre><code class="python">for chain_id in sequence['chain_id'].unique():
    print('\nChain ID: %s' % chain_id)
    print(''.join(sequence.loc[sequence['chain_id'] == chain_id, 'residue_name']))
</code></pre>

<pre><code>Chain ID: A
SLEPEPWFFKNLSRKDAERQLLAPGNTHGSFLIRESESTAGSFSLSVRDFDQGEVVKHYKIRNLDNGGFYISPRITFPGLHELVRHYT

Chain ID: B
SVSSVPTKLEVVAATPTSLLISWDAPAVTVVYYLITYGETGSPWPGGQAFEVPGSKSTATISGLKPGVDYTITVYAHRSSYGYSENPISINYRT
</code></pre>
<h2 id="wrapping-it-up-saving-pdb-structures">Wrapping it up - Saving PDB structures</h2>
<p>Finally, let's talk about how to get the PDB structures out of the DataFrame format back into the beloved .pdb format.</p>
<p>Let's say we loaded a PDB structure, removed it from it's hydrogens:</p>
<pre><code class="python">from biopandas.pdb import PandasPdb
ppdb = PandasPdb().read_pdb('./data/3eiy.pdb.gz')
ppdb.df['ATOM'] = ppdb.df['ATOM'][ppdb.df['ATOM']['element_symbol'] != 'H']
</code></pre>

<p>[File link: <a href="https://github.com/rasbt/biopandas/blob/master/docs/sources/tutorials/data/3eiy.pdb.gz?raw=true">3eiy.pdb.gz</a>]</p>
<p>We can save the file using the <a href="../api_subpackages/biopandas.pdb#pandaspdbto_pdb"><code>PandasPdb.to_pdb</code></a> method:</p>
<pre><code class="python">ppdb.to_pdb(path='./data/3eiy_stripped.pdb', 
            records=None, 
            gz=False, 
            append_newline=True)
</code></pre>

<p>[File link: <a href="https://raw.githubusercontent.com/rasbt/biopandas/master/docs/sources/tutorials/data/3eiy_stripped.pdb">3eiy_stripped.pdb</a>]</p>
<p>By default, all records (that is, 'ATOM', 'HETATM', 'OTHERS', 'ANISOU') are written if we set <code>records=None</code>. Alternatively, let's say we want to get rid of the 'ANISOU' entries and produce a compressed gzip archive of our PDB structure:</p>
<pre><code class="python">ppdb.to_pdb(path='./data/3eiy_stripped.pdb.gz', 
            records=['ATOM', 'HETATM', 'OTHERS'], 
            gz=True, 
            append_newline=True)
</code></pre>

<p>[File link: <a href="https://github.com/rasbt/biopandas/blob/master/docs/sources/tutorials/data/3eiy_stripped.pdb.gz?raw=true">3eiy_stripped.pdb.gz</a>]</p></div>
        </div>

        <footer class="col-md-12">
            <hr>
                <p>Copyright &copy; 2015-2018 <a href="http://sebastianraschka.com">Sebastian Raschka</a></p>
            <p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script>
            var base_url = "../..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
        <script src="../../js/base.js" defer></script>
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML" defer></script>
        <script src="../../mathjaxhelper.js" defer></script>
        <script src="../../search/main.js" defer></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                <h4 class="modal-title" id="exampleModalLabel">Search</h4>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form role="form">
                    <div class="form-group">
                        <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="Keyboard Shortcuts Modal" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                <h4 class="modal-title" id="exampleModalLabel">Keyboard Shortcuts</h4>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
